docs/spec.md

```diff
@@ -0,0 +1,2507 @@
+```
+
+### Undefine
+
+* 🔷 **removing aliases**
+```
+undefine alias p, q, r from person;
+undefine alias marriage:p, marriage:q, marriage:r from marriage:spouse;
+```
+
+### Redefine 
+
+* 🔷 **changing primary label**
+```
+redefine person label animal;
+redefine marriage:spouse label marriage:super_spouse;
+```
+
+# Pattern matching language
+
+This section first describes the pattern matching language of TypeDB, which relies on _variables_.
+
+## Basics: Patterns, variables, concept rows, satisfaction
+
+### (Theory) Statements and pattern
+
+* **statements:** syntactic units of TypeQL (see Glossary)
+* **patterns:** collection of statements, combined with logical connectives.
+
+### (Feature) Pattern operations
+
+#### **Case AND_PATT**
+
+* 🔷 _Conjunction_ `PATT1; PATT2` 
+  > We match "`PATT1` **and** `PATT2`" 
+
+  * _Note: terminating `;`_  Any `PATT` is terminated with a `;` (we could read this as "and true") 
+
+#### **Case OR_PATT**
+
+* 🔷 _Disjunction_ `PATT1 or PATT2` 
+  > We **either** match `PATT1` or `PATT2`
+
+  _Note_ This extends to $k$ patterns chained with `or`.
+
+#### **Case NOT_PATT**
+
+* 🔷 _Negation_. `not { PATT }` 
+  > We ensure `PATT` has **no match**. 
+
+  _Note_ `PATT` may have **quantified variables**
+
+#### **Case TRY_PATT**
+
+* 🔷 _Optional pattern_ `try { PATT }` 
+  > We **optionally** match `PATT` if possible
+
+  * For most purposes, this pattern is equivalent to `{ PATT } or { not { PATT }; }` **except** for the designation of **optional variables**
+
+_Terminology_ What's inside `{ ... }` is called a **block**
+
+
+### (Theory) Pattern branches
+
+A **disjunctive normal (DNF) of a pattern `PATT`** is a pattern obtained by
+
+* First unfolding all `try` blocks,
+* Then recursively applying transformations of the form 
+    ```
+    <PATT>; { <PATT_A> } or {<PATT_B> };
+    --transforms to-->
+    { <PATT>; <PATT_A>; } or { <PATT>; <PATT_B>; };
+    ```
+
+The resulting DNF of `PATT` will itself be a pattern of the form:
+```
+{ <BRANCH_1> } or { <BRANCH_2> } or ...
+```
+
+**Important**. The patterns `<BRANCH_i>` are _unique_ up to re-ordering them. We call them the **branches** of `PATT`.
+
+
+### (Theory) Variables and bindings
+
+Variables appear in statements. They fall in different categories, which can be recognized as follows.
+
+* **Syntax**: variables in patterns `PATT` are `$`-prefixed labels
+  * _Examples_: `$x`, `$y`, `$person`
+
+#### **Variable categories**
+
+* **Categories** indicate what type of concept a variable can hold. In a pattern `PATT`, any variable will belong to one of four _categories_ based on the following rules.
+
+  * **Type variables** (**tvar**, uppercase convention in this spec)
+    * Any variable used in a type position in a statement
+
+  * **Value variables** (**vvar**, lowercase convention in this spec)
+    * Any variable which are typed with non-comparable attribute types is a value variables (**#BDD**)
```

Comment:
I understood this, but it could be clearer. Is "non-comparable" previously defined?